AdminSessionsManager and AdminSessionInfo - strtotime issue fix for admin login PR

kanhaiya5590 opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
 - TypeError: strtotime() expects parameter 1 to be string, int given ...security/Model/AdminSessionsManager.php:338 
 - TypeError: strtotime() expects parameter 1 to be string, null given in ...security/Model/AdminSessionInfo.php:136

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#34415
2. Fixes https://github.com/magento/magento2/issues/34531

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

kanhaiya5590 committed:
main.CRITICAL: TypeError: strtotime() expects parameter 1 to be string, null

main.CRITICAL: TypeError: strtotime() expects parameter 1 to be string, null given in ...security/Model/AdminSessionInfo.php:136

kanhaiya5590 committed:
strtotime() expects parameter 1 to be string, int ...security/Model/AdminSessionsManager.php(338)

TypeError: strtotime() expects parameter 1 to be string, int given
#28 /var/www/share/meevo.io/releases/66/vendor/magento/module-security/Model/AdminSessionsManager.php(338):

kanhaiya5590 commented:
@magento give me 2.4-develop instance

Den4ik commented on app/code/Magento/Security/Model/AdminSessionInfo.php:
```diff
@@ -133,7 +133,11 @@ public function isSessionExpired()
-            $lastUpdatedTime = strtotime($lastUpdatedTime);
+            if($lastUpdatedTime === null) {
+                $lastUpdatedTime = 0;
+            }else {
```
```suggestion
            } else {
```

Den4ik committed:
Update app/code/Magento/Security/Model/AdminSessionInfo.php

Den4ik commented:
@magento run all tests

kanhaiya5590 committed:
update spaces

kanhaiya5590 committed:
updated space

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

kanhaiya5590 commented:
@magento run all tests

Den4ik committed:
Fix static tests

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik committed:
Fix static tests

Signed-off-by: Denis Kopylov <dkopylov@magenius.team>

Den4ik commented:
@magento run all tests

hostep commented:
Watch out that there are 2 different PR's trying to fix the same issue:
- this one
- and: https://github.com/magento/magento2/pull/34520

Only one of both should be chosen, right?

johnulecan commented:
Hey @hostep . I guess only one PR will be accepted. However, I think there is no need to initialize a timestamp with 0 and also there is no need to add additional conditions. The problem lies in AdminSessionInfo.php isSessionExpired() - wreturn bool - method. If the session is expired, getUpdatedAt returns null and strtotime expects string. This fixes the whole problem: https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34520.diff

Den4ik commented:
Yep, only one PR could be accepted

ihor-sviziev commented on app/code/Magento/Security/Model/AdminSessionInfo.php:
```diff
@@ -133,7 +134,11 @@ public function isSessionExpired()
         $lastUpdatedTime = $this->getUpdatedAt();
         if (!is_numeric($lastUpdatedTime)) {
-            $lastUpdatedTime = strtotime($lastUpdatedTime);
+            if ($lastUpdatedTime === null) {
+                $lastUpdatedTime = 0;
+            } else {
+                $lastUpdatedTime = strtotime($lastUpdatedTime);
+            }
```
```suggestion
            $lastUpdatedTime = $lastUpdatedTime === null ? 0 : strtotime($lastUpdatedTime);
```

ihor-sviziev commented on app/code/Magento/Security/Model/AdminSessionInfo.php:
```diff
@@ -133,7 +134,11 @@ public function isSessionExpired()
+            }
         }
 
         return $lastUpdatedTime <= ($currentTime - $lifetime) ? true : false;
```
```suggestion
        return $lastUpdatedTime <= ($currentTime - $lifetime);
```

kanhaiya5590 replied:
this remove make no sense but as return will be either true or false, to optimize point of view its good to remove

ihor-sviziev reviewed (changes requested):
Hi @kanhaiya5590,
Could you apply suggested changes and cover your changes with unit tests?

kanhaiya5590 committed:
Update app/code/Magento/Security/Model/AdminSessionInfo.php

Co-authored-by: Ihor Sviziev <ihor-sviziev@users.noreply.github.com>

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

kanhaiya5590 committed:
Update app/code/Magento/Security/Model/AdminSessionInfo.php

Co-authored-by: Ihor Sviziev <ihor-sviziev@users.noreply.github.com>

kanhaiya5590 commented:
Hi @hostep @johnulecan @ihor-sviziev @Den4ik ,

Can you help to add Test Cases for changes.

or @magento-engcom-team  help for unit test

kanhaiya5590 committed:
Update AdminSessionInfoTest.php

unit test case added - https://github.com/kanhaiya5590/magento2/commit/32a742eb5b00249f757df079ea373c975ed8d1de

kanhaiya5590 committed:
Update AdminSessionsManagerTest.php

test case https://github.com/Tjitse-E/magento2/commit/32a742eb5b00249f757df079ea373c975ed8d1de# thanks to @Tjitse-E - https://github.com/Tjitse-E/

kanhaiya5590 commented:
@magento run all tests

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

kanhaiya5590 commented:
@magento run Unit Tests

kanhaiya5590 commented:
> @kanhaiya5590 i've created 2 unit tests, maybe you can use them for your PR: [Tjitse-E@32a742e](https://github.com/Tjitse-E/magento2/commit/32a742eb5b00249f757df079ea373c975ed8d1de)

@Tjitse-E  Thanks for Unit tests - added in PR

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @kanhaiya5590,
Now the changes look good to me.
Could you fix failing static tests and I'll approve your PR?

kanhaiya5590 committed:
Update AdminSessionsManagerTest.php

 64 | WARNING | Missing PHP DocBlock for class property.

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

kanhaiya5590 committed:
Update AdminSessionsManagerTest.php

 Line exceeds 120 characters; contains 137 characters

kanhaiya5590 commented:
@magento run Static Tests

ihor-sviziev commented:
@kanhaiya5590 Great job!
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9336](https://jira.corp.magento.com/browse/ENGCOM-9336) has been created to process this Pull Request

lytesaber commented:
I see this is being merged into 2.4-develop branch however this problem also the affects 2.3.7-p2 release are we going to get a patch for the 2.3.x release line as well? Currently we're having to ask clients to clear their cookies to be able to access the admin panel of their stores due to this error.

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@sidolov, how can we convert this PR to a patch to 2.3.7-p2?

kanhaiya5590 commented:
HI @ihor-sviziev 

Every pull-request on GH can be downloaded as a beautiful mail-patch, just by appending ".patch" to the PR URL

https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34514.patch 

This will generate a mail-formatted patch file, that is a little different from an usual patch file: It has e-mail metadata.

After downloaded, apply the result patch using:
`$ git am 34514.patch`

Hope this is what patch you looking for.

or PR diff -> https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34514.diff

Den4ik commented:
> @sidolov, how can we convert this PR to a patch to 2.3.7-p2?

Similar as cookie patch. I made PR manually. @kanhaiya5590 Could you prepare PR to 2.3-release branch?

magento-engcom-team commented:
Hi @Den4ik, thank you for the review.
[ENGCOM-9336](https://jira.corp.magento.com/browse/ENGCOM-9336) has been created to process this Pull Request

kanhaiya5590 commented:
Hi @Den4ik 
Created  a PR for Magento v2.3 https://github.com/magento/magento2/pull/34606

ihor-sviziev commented:
@kanhaiya5590, yeah, I know, the question was about creating the quality patch for magento 2.3.7-p2

ihor-sviziev commented:
Hi @kanhaiya5590,
It looks like this issue isn't reproducing on clean Magento installation. Could you provide some steps to reproduce it?

DannyDaemonic commented:
I just ran into issue #34415. Applying this PR (#34514) resolved the problem.

@ihor-sviziev I apologize if this isn't productive, as I can't say the exact steps, but it seems it was at least partially triggered by logging in from a different system and ip address without first logging out of my previous session. When I went back to the system I was previously logged in to, I ran into the issue.

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento latest installed


**Manual testing scenario:**

1. Update the code as shown in below screenshot.
<img width="1440" alt="Screenshot 2021-12-27 at 2 29 44 PM" src="https://user-images.githubusercontent.com/51679138/147460820-ea9f13e1-b63d-4aeb-83de-22f5af779bc3.png">

2. Try to login as an Admin from UI.


**Before:** :heavy_multiplication_x: Getting an typeError:

<img width="1440" alt="Screenshot 2021-12-27 at 2 29 55 PM" src="https://user-images.githubusercontent.com/51679138/147460945-f2dc8c91-ad53-4eff-bcd8-69e45d436a8a.png">


**After:** :heavy_check_mark: Able to see the values in the response 


Since it is a specific case were code changes are involved, no additional regression is required on this!

kanhaiya5590 committed:
Merge branch '2.4-develop' into 2.4-develop

kanhaiya5590 commented:
Hope this PR is ready to merge.

FYi - @ihor-sviziev @engcom-Alfa @Den4ik
